tetraelf-hierarchy/hierarchy-portal/impl/src/java/org/sakaiproject/portal/impl/HierarchySiteNeighbourhoodService.java

```diff
@@ -177,17 +179,50 @@ public String parseSiteAlias(String alias) {
 	}
 
 	@Override
-	public String getRedirect(String siteId) {
-		if (siteId != null) {
-			PortalNode node = portalHierarchyService.getNodeByUrl(siteId);
+	public Redirect getRedirect(String siteId) {
+		if (StringUtils.isNotBlank(siteId)) {
+			Redirect redirect = new Redirect();
+			PortalNode node;
+			String trimmedPath = "";
+			do{
+				node = portalHierarchyService.getNodeByUrl(siteId);
+				//check for the redirect of subsites
+				if(node == null){
+					//get the index of last occurrence of Separator in the path
+					int index = siteId.lastIndexOf(PortalHierarchyService.SEPARATOR);
+					if(index > 0){
+						String firstPart = siteId.substring(0, index);
+						String tempString = PortalHierarchyService.SEPARATOR + siteId.substring(index + 1);
+						//save the trimmed path if there is a subsite redirect
+						trimmedPath = (StringUtils.isBlank(trimmedPath)) ? tempString : (tempString + trimmedPath) ;
+						siteId = firstPart;
+					}
+				}
+				else{
+					//node found skip the loop
+					break;
+				}
+			}while(StringUtils.countMatches(siteId, PortalHierarchyService.SEPARATOR) > 1);
```

Comment:
This is a risky pattern as if there is a problem in the code we can get runaway threads, it's much safer to have a known loop.